### PR TITLE
ENH: add str.removeprefix

### LINKED ISSUE #36944: ENH: add string method remove prefix and suffix, python 3.9
Opening this issue for discussion:

Since python 3.9 is out, and we have the new [string methods](https://www.python.org/dev/peps/pep-0616/) `removeprefix` and `removesuffix`, it would be nice to add them to the pandas string methods as well
```python
In [2]: import pandas as pd

In [3]: df = pd.DataFrame({'A': ['str_string1', 'str_string2', 'str_string3']})

In [4]: print(df)
             A
0  str_string1
1  str_string2
2  str_string3

In [5]: df['A'].str.removeprefix('str_')
Out[5]: 
0    string1
1    string2
2    string3
Name: A, dtype: object
```
An argument not to add this is that it's pretty easily to achieve with `str.split`:
```python
df['A'].str.split('_').str[-1]
```

### PR DESCRIPTION
**Pull request by erfannariman**
- [x] closes #36944
- [ ] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/strings/accessor.py [hunk 1]
```diff
@@ -2885,6 +2885,14 @@ def casefold(self):
         "isdecimal", docstring=_shared_docs["ismethods"] % _doc_args["isdecimal"]
     )
 
+    @forbid_nonstring_types(["bytes"])
+    def removeprefix(self, prefix=None):
+        if not prefix:
+            return self._data
```
Comment (jreback) [inline pandas/core/strings/accessor.py:2891]: hmm do we copy in other cases?

File: pandas/core/strings/accessor.py [hunk 2]
```diff
@@ -2885,6 +2885,14 @@ def casefold(self):
         "isdecimal", docstring=_shared_docs["ismethods"] % _doc_args["isdecimal"]
     )
 
+    @forbid_nonstring_types(["bytes"])
+    def removeprefix(self, prefix=None):
```
Comment (jreback) [inline pandas/core/strings/accessor.py:2889]: can you add an example in the doc string

File: pandas/core/strings/object_array.py [hunk 3]
```diff
@@ -430,3 +430,6 @@ def _str_lstrip(self, to_strip=None):
 
     def _str_rstrip(self, to_strip=None):
         return self._str_map(lambda x: x.rstrip(to_strip))
+
+    def _str_removeprefix(self, prefix=None):
+        return self._str_map(lambda x: x.removeprefix(prefix))
```
Comment (jreback) [inline pandas/core/strings/object_array.py:435]: this only exists in 3.9 right? is there an impl that can work for other versions as well

File: pandas/tests/test_strings.py [hunk 4]
```diff
@@ -307,8 +309,18 @@ def test_api_for_categorical(self, any_string_method):
 
         method_name, args, kwargs = any_string_method
 
-        result = getattr(c.str, method_name)(*args, **kwargs)
-        expected = getattr(s.str, method_name)(*args, **kwargs)
+        # we expect an AttributeError when str.removeprefix is tested on <= 3.9
```
Comment (jreback) [inline pandas/tests/test_strings.py:312]: we want this to exist for all versions.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/strings/object_array.py (hunk 3 above)
Comment (davidwales) [reply pandas/core/strings/object_array.py:435]: This is the spec from [PEP 616](https://www.python.org/dev/peps/pep-0616/#specification):

``` python
def removeprefix(self: str, prefix: str, /) -> str:
    if self.startswith(prefix):
        return self[len(prefix):]
    else:
        return self[:]

def removesuffix(self: str, suffix: str, /) -> str:
    # suffix='' should not call self[:-0].
    if suffix and self.endswith(suffix):
        return self[:-len(suffix)]
    else:
        return self[:]
```

**Comment by jreback** [comment]
@erfannariman can you merge master and fixup

**Comment by erfannariman** [comment]
> @erfannariman can you merge master and fixup

Yes, will do this weekend.

**Comment by WillAyd** [comment]
Will also want to add an entry to doc/source/reference/series.rst to make sure this renders in the docs

**Comment by jreback** [comment]
yep, pls add to here: https://pandas.pydata.org/pandas-docs/dev/user_guide/text.html#method-summary, optionally add add an example near strip

**Comment by simonjayhawkins** [comment]
@erfannariman can you merge master and get tests to pass. also can you add a benchmark.

Is there a benefit to using the builtin if py39 else using the fallback

and also for the fallback is there any benefit to having two maps followed by a np.where over just using the fallback function (len(prefix) is invariant so the mapper function can be defined in a closure)

This does not close the issue? need to add removesuffix also?

This should also work for StringArray. can you use `any_string_dtype` in the tests

**Comment by simonjayhawkins** [comment]
Thanks @erfannariman for the PR. closing as stale. ping if you want to continue.

**Pull request closed without merging**